libimage/runtime.go

```diff
@@ -574,23 +574,30 @@ func (r *Runtime) ListImages(ctx context.Context, names []string, options *ListI
 		options = &ListImagesOptions{}
 	}
 
-	var images []*Image
-	if len(names) > 0 {
-		for _, name := range names {
-			image, _, err := r.LookupImage(name, nil)
-			if err != nil {
-				return nil, err
-			}
-			images = append(images, image)
-		}
-	} else {
-		storageImages, err := r.store.Images()
+	lookupImagesMap := make(map[string]struct{}, 0)
+	for _, name := range names {
+		image, _, err := r.LookupImage(name, nil)
 		if err != nil {
 			return nil, err
 		}
-		for i := range storageImages {
-			images = append(images, r.storageToImage(&storageImages[i], nil))
+		lookupImagesMap[image.ID()] = struct{}{}
+	}
+
+	storageListResult, err := r.store.MultiList(
+		storage.MultiListOptions{Images: true, Layers: true},
```

Comment:
Some systems have many images and very many layers; it would be better to only ask for layers if we are actually going to use them.

Reply:
All layers are used to create a `layerTree`. The same approach was used in the previous version.

Reply:
No, the current version does not create `layerTree` if `!options.SetListData` and the filters don‘t require the tree.

Many call sites are like that, so we shouldn’t _always_ add this overhead.